src/Symfony/Component/Messenger/Middleware/HandleMessageInNewTransactionMiddleware.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Middleware;
+
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\EnvelopeAwareInterface;
+use Symfony\Component\Messenger\Exception\MessageHandlingException;
+use Symfony\Component\Messenger\Middleware\Configuration\Transaction;
+
+/**
+ * Allow to configure messages to be handled in a new Doctrine transaction if using
+ * the DoctrineTransactionMiddleware. This middleware should be used before DoctrineTransactionMiddleware.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class HandleMessageInNewTransactionMiddleware implements MiddlewareInterface, EnvelopeAwareInterface
+{
+    /**
+     * @var array A queue of messages and callables
+     */
+    private $queue = array();
+
+    /**
+     * @var bool Indicate if we are running the middleware or not. Ie, are we called inside a message handler?
+     */
+    private $insideMessageHandler = false;
+
+    /**
+     * @param Envelope $envelope
+     */
+    public function handle($envelope, callable $next)
+    {
+        if (null !== $envelope->get(Transaction::class)) {
+            if (!$this->insideMessageHandler) {
+                throw new \LogicException('We have to use the transaction in the context of a message handler');
+            }
+            $this->queue[] = array('envelope' => $envelope, 'callable' => $next);
+
+            return;
+        }
+
+        if ($this->insideMessageHandler) {
+            /*
+             * If come inside a second message handler, just continue as normal. We should not
+             * run the stored messages.
+             */
+            return $next($envelope);
+        }
+
+        $this->insideMessageHandler = true;
```

Comment:
So, this will only work when dispatching from the same bus, or by using the same middleware instance across event and command buses, right? 
If that's intended, then I guess the service definition should not be abstract, otherwise the messenger pass will create distinct instances.